File: solution/202HappyNumber/202HappyNumber.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Solution {
public:
  bool isHappy(int n) {
    unordered_map<int, bool> m;
    m[n] = true;
    while (n != 1) {
      int s = 0;
      while (n != 0) {
        int d = n % 10;
        n /= 10;
        s += d * d;
      }
      // cout << s << endl;
      if (m.find(s) != m.end()) {
        return false;
      } else {
        m[s] = true;
      }
      n = s;
    }
    return true;
  }
};

int main() {
  Solution s;
  cout << s.isHappy(19) << endl;
  cout << s.isHappy(0) << endl;
  cout << s.isHappy(1) << endl;
  cout << s.isHappy(1239223) << endl;
}

File: src/202HappyNumber/202HappyNumber.rs
impl Solution {
    pub fn is_happy(n: i32) -> bool {
        let mut s = std::collections::HashSet::new();
        let mut m = n;
        while !s.contains(&m) {
            s.insert(m);
            m = Self::sum_of_square(m);
            if m == 1 {
                return true;
            }
        }
        false
    }
    fn sum_of_square(n: i32) -> i32{
        let mut m = n;
        let mut sum = 0;
        while m!= 0 {
            let d = m%10;
            m = m/10;
            sum += d*d;
        }
        sum
    }
}
